<a href="https://colab.research.google.com/github/varun-beep/LLM_Lab/blob/main/Varun_555_Lab2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install transformers

In [2]:
from transformers import AutoTokenizer, AutoModelForCausalLM, AutoModelForSeq2SeqLM
import torch

In [3]:
gpt2_tokenizer = AutoTokenizer.from_pretrained("gpt2")
gpt2_model = AutoModelForCausalLM.from_pretrained("gpt2")

bart_tokenizer = AutoTokenizer.from_pretrained("facebook/bart-base")
bart_model = AutoModelForSeq2SeqLM.from_pretrained("facebook/bart-base")

flan_tokenizer = AutoTokenizer.from_pretrained("google/flan-t5-base")
flan_model = AutoModelForSeq2SeqLM.from_pretrained("google/flan-t5-base")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.72k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/558M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/2.54k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.40k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/990M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

In [4]:
prompts = [
    "A lonely robot discovers a hidden message from the past.",
    "In a city where dreams are bought and sold, a boy tries to steal one.",
    "An old woman finds a book that tells the future—and it shows her death."
]

In [5]:
def generate_story(prompt, model, tokenizer, max_len=200):
    inputs = tokenizer(prompt, return_tensors="pt")
    if isinstance(model, AutoModelForCausalLM):
        output = model.generate(
            **inputs,
            max_length=max_len,
            temperature=0.8,
            top_k=50,
            top_p=0.95,
            do_sample=True,
            pad_token_id=tokenizer.eos_token_id
        )
    else:
        output = model.generate(
            **inputs,
            max_length=max_len,
            temperature=0.8,
            top_k=50,
            top_p=0.95,
            do_sample=True
        )
    return tokenizer.decode(output[0], skip_special_tokens=True)

In [6]:
for prompt in prompts:
    print(f"\n🔹 Prompt: {prompt}\n")

    print("📘 GPT-2 Story:")
    print(generate_story(prompt, gpt2_model, gpt2_tokenizer))

    print("\n📙 BART Story:")
    print(generate_story(prompt, bart_model, bart_tokenizer))

    print("\n📗 FLAN-T5 Story:")
    print(generate_story(prompt, flan_model, flan_tokenizer))

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.



🔹 Prompt: A lonely robot discovers a hidden message from the past.

📘 GPT-2 Story:
A lonely robot discovers a hidden message from the past. The world is not finished. We are going to save the world. "

— John Carter, "The Martian"

While the android has been seen on the Martian surface over the years, this is the first time she has encountered it and has been known as a Martian. In reality, it is not her that is seen as human, but her "husband" (a.k.a. "Shepherd"). Her husband is a scientist named James Mars (later a member of the Earth Liberation Front). She is not really human, but rather a "mama robot" with a "cute body". She is a "smart, human, and gentle" robot and often uses her voice to "talk to" other robot people.

She is the first robot to be seen on the planet before the fall of the Martian city. She is also the first robot to be seen on Mars.

She is

📙 BART Story:
A lonely robot discovers a hidden message from the past.

📗 FLAN-T5 Story:


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


The robot is a perfect example of the art of self-discovery.

🔹 Prompt: In a city where dreams are bought and sold, a boy tries to steal one.

📘 GPT-2 Story:
In a city where dreams are bought and sold, a boy tries to steal one. And the man tries to steal something from his own home. And the man is forced to sell something from the house he was born in.

A few weeks ago, a 17-year-old boy was walking his dog when he was told to stop. The boy did not stop, and he was taken off the dog, taken to the hospital, and forced to leave.

He was then forced to move back in with his mother and brother.

The boy's mother and brother are both in jail. The boy's father is in jail, too.

So, what happened to the boy who was told to stop and that he was being sold?

The police asked the boy, "How can you be selling a child in such a situation?"

The boy responded, "I don't care about your child, you can come and play, you can sit,

📙 BART Story:
In a city where dreams are bought and sold, a boy tries t

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


He then steals a doll and then tells his parents he wants to take one.

🔹 Prompt: An old woman finds a book that tells the future—and it shows her death.

📘 GPT-2 Story:
An old woman finds a book that tells the future—and it shows her death.

A few days later, her friend, a painter, finds a book that tells the future—and it shows her death.

A few days later, a friend finds a book that tells the future—and it shows her death.

The same day, someone has a book of poems that tells the story of their friend, a painter.

A few days later, someone has a book that tells the story of their friend, a painter.

And then somebody finds a book about a woman who died.

A few days later, someone finds a book about a woman who died.

And then somebody finds a book about a woman who died.

People's names are changed from one person to another.

A few days later, someone finds a book about a woman who died.

A few days later, someone finds a book about a

📙 BART Story:
An old woman finds a book that t

In [8]:
# 📌 Install nltk + rouge_score
!pip install nltk rouge-score --quiet
import nltk
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from rouge_score import rouge_scorer

nltk.download('punkt')
nltk.download('punkt_tab')
smoothie = SmoothingFunction().method4
rouge = rouge_scorer.RougeScorer(['rouge1', 'rougeL'], use_stemmer=True)

# ✅ Reference story for testing
reference = "The robot found an old recording hidden under debris. It was a message from its creator, explaining its purpose. The robot now had meaning."

# ✅ Evaluate model outputs for 1st prompt
prompt = prompts[0]
gpt2_story = generate_story(prompt, gpt2_model, gpt2_tokenizer)
bart_story = generate_story(prompt, bart_model, bart_tokenizer)
flan_story = generate_story(prompt, flan_model, flan_tokenizer)

# Tokenize reference and predictions
ref_tokens = nltk.word_tokenize(reference.lower())
predictions = {
    "GPT-2": gpt2_story,
    "BART": bart_story,
    "FLAN-T5": flan_story
}

# ✅ Print BLEU & ROUGE scores
for model_name, story in predictions.items():
    pred_tokens = nltk.word_tokenize(story.lower())
    bleu = sentence_bleu([ref_tokens], pred_tokens, smoothing_function=smoothie)
    rouge_scores = rouge.score(reference, story)

    print(f"\n📊 Evaluation for {model_name}:")
    print(f"🟦 BLEU Score: {bleu:.4f}")
    print(f"🟥 ROUGE-1: {rouge_scores['rouge1'].fmeasure:.4f}")
    print(f"🟩 ROUGE-L: {rouge_scores['rougeL'].fmeasure:.4f}")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.



📊 Evaluation for GPT-2:
🟦 BLEU Score: 0.0065
🟥 ROUGE-1: 0.0860
🟩 ROUGE-L: 0.0538

📊 Evaluation for BART:
🟦 BLEU Score: 0.0151
🟥 ROUGE-1: 0.3529
🟩 ROUGE-L: 0.2941

📊 Evaluation for FLAN-T5:
🟦 BLEU Score: 0.0133
🟥 ROUGE-1: 0.2927
🟩 ROUGE-L: 0.1463


In [9]:
# 📌 Install Gradio
!pip install gradio --quiet

import gradio as gr

# ✅ Function for Gradio
def generate_flan_story(prompt, temperature=0.8, max_len=300):
    inputs = flan_tokenizer(prompt, return_tensors="pt")
    output = flan_model.generate(
        **inputs,
        max_length=max_len,
        temperature=temperature,
        top_k=50,
        top_p=0.95,
        do_sample=True
    )
    return flan_tokenizer.decode(output[0], skip_special_tokens=True)

# ✅ Gradio Interface
gr.Interface(
    fn=generate_flan_story,
    inputs=[
        gr.Textbox(label="Enter your story prompt here"),
        gr.Slider(minimum=0.1, maximum=1.5, value=0.8, label="Temperature"),
        gr.Slider(minimum=100, maximum=500, value=300, label="Max Length")
    ],
    outputs=gr.Textbox(label="Generated Story"),
    title="📝 FLAN-T5 Story Generator",
    description="Enter a creative prompt and generate a story using FLAN-T5!"
).launch()


It looks like you are running Gradio on a hosted a Jupyter notebook. For the Gradio app to work, sharing must be enabled. Automatically setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://e15292ae80ea0b9a3e.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


| **Aspect**             | **GPT-2**                                   | **BART**                        | **FLAN-T5**                          |
| ---------------------- | ------------------------------------------- | ------------------------------- | ------------------------------------ |
| **Architecture**       | Decoder-only (Causal)                       | Encoder-Decoder                 | Encoder-Decoder (Instruction-tuned)  |
| **Training Objective** | Predict next token (left-to-right)          | Denoising autoencoder (Seq2Seq) | Instruction-based generation         |
| **Output Style**       | Imaginative but inconsistent and repetitive | Very concise and literal        | Short, vague, but logical            |
| **Story Quality**      | Sometimes derails into unrelated loops      | Mostly repeats the prompt only  | Slightly better attempt at structure |
| **Creativity**         | 🔸 Medium-High                              | 🔸 Low                          | 🔸 Medium                            |
| **Coherence**          | ⚠️ Weak                                     | ✅ Strong                        | ⚠️ Moderate                          |
| **BLEU Score**         | 0.0065                                      | 0.0151 ✅                        | 0.0133                               |
| **ROUGE-1**            | 0.0860                                      | 0.3529 ✅                        | 0.2927                               |
| **ROUGE-L**            | 0.0538                                      | 0.2941 ✅                        | 0.1463                               |
| **Best Use Case**      | Open-ended story generation                 | Summarization or paraphrasing   | Instruction-following & QA           |
| **Limitation**         | Repetition & hallucination                  | Lacks creativity                | Very brief output                    |



Among the three models, BART performed best in coherence and fluency as reflected by the highest BLEU and ROUGE scores, even though it lacked creativity. GPT-2 was the most imaginative but often generated erratic or repetitive stories. FLAN-T5 was concise and sometimes logical, but failed to generate detailed content beyond a sentence or two.

Justification:
Balanced performance in coherence and evaluation scores
Better control over structure due to its encoder-decoder format
Ideal for short-form, logical, and grammatically correct outputs